i6_models/primitives/mixup.py

```diff
@@ -0,0 +1,153 @@
+        if not self.filled and self.pos == 0:
+            return None
+        else:
+            end_idx = self.buffer_size if self.filled else self.pos
+            max_end_idx = end_idx - t_dim
+
+            start_indicies = torch.randint(
+                high=max_end_idx, size=(b_dim, max_num_mixup)
+            )  # [B, M] (M denotes maximum of num_mixup over the batch)
+            start_indicies_flat = torch.masked_select(
+                start_indicies, n_mask
+            )  # [B, M'] (M' denotes sum of num_mixup over the batch)
+
+            idx = torch.arange(t_dim)
+            idx = torch.unsqueeze(idx, dim=-1) + start_indicies_flat  # [T, M']
+            mixup_values = self.cache[idx]  # [T, M', F]
+            return mixup_values
+
+
+class Mixup:
+    """
+    Implement the Mixup data augmentation method
+    C.f. https://arxiv.org/abs/1710.09412
+
+    The code is partly based on Albert's implementation in returnn front-end
+    C.f. https://github.com/rwth-i6/i6_experiments/blob/main/users/zeyer/returnn/models/rf_mixup.py
+    """
+
+    def __init__(self, feature_dim: int, cfg: MixupConfig):
+        self.apply_prob = cfg.apply_prob
+        self.lambda_min = cfg.lambda_min
+        self.lambda_max = cfg.lambda_max
+        self.max_num_mix = cfg.max_num_mix
+        self.buffer_size = cfg.buffer_size
+        self.feature_dim = feature_dim
+
+        self.feature_buffer = FeatureBuffer(self.buffer_size, feature_dim)
+
+    def __call__(self, input: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
+        assert (
+            input.size()[-1] == self.feature_dim
+        ), "the given feature dimension does not match input feature dimension"
+        input_ = self._maybe_apply_mixup(input, sequence_mask)
+        self._append_to_buffer(input, sequence_mask)
+        return input_
+
+    def _append_to_buffer(self, input: torch.Tensor, sequence_mask: torch.Tensor):
+        _, _, f_dim = input.size()
+        # mask out the padded frames before append it to feature_buffer
+        input_flat = torch.masked_select(input, sequence_mask.unsqueeze(-1)).view(-1, f_dim)
+        self.feature_buffer.append(input_flat)
+
+    def _maybe_apply_mixup(self, input: torch.Tensor, sequence_mask: torch.Tensor) -> torch.Tensor:
+        if random.random() >= self.apply_prob:
+            return input
+
+        b_dim, t_dim, f_dim = input.size()  # [B, T, F]
+
+        max_end_idx = self.feature_buffer.buffer_size if self.feature_buffer.filled else self.feature_buffer.pos
+        if t_dim > max_end_idx:
+            return input
+
+        num_mixup = torch.randint(low=1, high=self.max_num_mix + 1, size=(b_dim,))  # [B]
+        max_num_mixup = max(num_mixup)  # [M] (M denotes maximum of num_mixup)
+
+        row_vector = torch.arange(0, max_num_mixup, 1)  # [M]
+        n_mask = torch.unsqueeze(num_mixup, dim=-1) > row_vector  # [B, M]
```

Comment:
I think the broadcasting is more clear this way:
```suggestion
        n_mask = row_vector[None, :] < num_mixup[:, None]  # [B, M]
```